### PR TITLE
rewrite comment_box as functional

### PR DESCRIPTION
**Pull request by goapunk**

### CODE REVIEW

File: adhocracy4/comments_async/static/comments_async/comment_box.jsx [hunk 1]
```diff
@@ -93,167 +87,140 @@ export default class CommentBox extends React.Component {
...
+    } else {
+      diff = { $unshift: [comment] }
+      newCommentCount++
+      setMainError(undefined)
+    }
+    setComments(update(comments, diff))
+    setCommentCount(newCommentCount)
+  }
+
+  function setReplyError (parentIndex, index, message) {
```
Comment (goapunk) [inline adhocracy4/comments_async/static/comments_async/comment_box.jsx:158]: I've added a bunch of convenient functions for the error handling as I found it hard to read

File: adhocracy4/comments_async/static/comments_async/comment_box.jsx [hunk 2]
```diff
@@ -93,167 +87,140 @@ export default class CommentBox extends React.Component {
...
-        flushSync(() => {
-          this.updateStateComment(
-            index,
-            parentIndex,
-            changed)
-        })
         this.updateStateComment(
           index,
-          parentIndex, {
+          parentIndex,
```
Comment (goapunk) [inline adhocracy4/comments_async/static/comments_async/comment_box.jsx:205]: merged the two setState calls into one here

**Review by phillimorland — APPROVED** [review]
thats all i've found so far

File: adhocracy4/comments_async/static/comments_async/comment_box.jsx [hunk 3]
```diff
@@ -264,341 +226,288 @@ export default class CommentBox extends React.Component {
...
-          </div>
-        </div>
-        <div className={this.state.loading ? 'a4-comments__loading' : 'd-none'}>
+            {props.withCategories
+              ? <FilterCategory
+                  translated={translated}
+                  filter={filter}
+                  filterDisplay={filterDisplay}
+                  onClickFilter={handleClickFilter}
+                  commentCategoryChoices={props.commentCategoryChoices}
```
[the remark this answers is not in the document]
Comment (goapunk) [reply adhocracy4/comments_async/static/comments_async/comment_box.jsx:468]: nice catch! fixed

**Review by khamui — COMMENTED** [review]
Very nice! It is working mostly fine.
What I didn't test, but it looks fine from code:

- Scroll Event listener
- Errorhandling

what i did test:
- post/edit/delete comments and subcomments
- TOS behavior (that seems to be handled in the other PR)
- sort/filter/search

File: adhocracy4/comments_async/static/comments_async/comment_box.jsx [hunk 4]
```diff
@@ -93,167 +82,140 @@ export default class CommentBox extends React.Component {
...
-          })
-        }
+        const newErrorMessage = (Object.values(xhr.responseJSON))[0]
+        setReplyError(parentIndex, undefined, newErrorMessage)
       })
   }
 
-  handleCommentModify (modifiedComment, index, parentIndex) {
-    const comments = this.state.comments
+  function handleCommentModify (modifiedComment, index, parentIndex) {
```
Comment (khamui) [inline adhocracy4/comments_async/static/comments_async/comment_box.jsx:191]: probably leftover `this.updateStateComment...` , `this` needs to be removed.

**Review by khamui — CHANGES_REQUESTED** [review]

File: adhocracy4/comments_async/static/comments_async/comment_box.jsx [hunk 5]
```diff
@@ -33,57 +32,47 @@ const translated = {
...
+  const [sort, setSort] = useState(props.useModeratorMarked ? 'mom' : 'new')
+  const [loading, setLoading] = useState(true)
+  const [loadingFilter, setLoadingFilter] = useState(false)
+  const [search, setSearch] = useState('')
+  const [anchoredCommentParentId, setAnchoredCommentParentId] = useState(0)
+  const [anchoredCommentFound, setAnchoredCommentFound] = useState(false)
+  const [hasCommentingPermission, setHasCommentingPermission] = useState(false)
+  const [wouldHaveCommentingPermission, setWouldHaveCommentingPermission] = useState(false)
+  const [projectIsPublic, setProjectIsPublic] = useState(false)
+  const [useTermsOfUse, setUseTermsOfUse] = useState(true)
```
[the remark this answers is not in the document]
Comment (goapunk) [reply adhocracy4/comments_async/static/comments_async/comment_box.jsx:55]: good point, I set it to false

**Review by khamui — APPROVED** [review]

**Comment by phillimorland** [comment]
this doesn't help but maybe more people will comment https://github.com/facebook/react/issues/24696

**Review by phillimorland — APPROVED** [review]
looking good.
tested the scroll and error and found (parts Kha didn't):
the auto scroll for anchored comments works 
errors work however the auto focus that browsers have to `role=alert` is not working however this seems also broken on dev. so nothing to do with this pr :)

**Pull request merged**